test/kad-dht.js

```diff
@@ -1,112 +1,692 @@
+            nodes[0].api.swarm.peers((err, peers) => {
+              if (err) {
+                return cb(err)
+              }
+              cb(null, peers.length === 3)
+            })
+          }
+
+          // Wait for peers to have each other on the DHT
+          waitFor(testDhtReady, {
+            name: 'dht peers ready',
+            timeout: 30 * 1000,
+            interval: 5 * 1000
+          }, done)
+        })
+      })
+
+      after(function (done) {
+        this.timeout(80 * 1000)
+        each(nodes, (node, cb) => {
+          node.stop(cb)
+        }, done)
+      })
+
+      it('should get from the network after being added', function (done) {
+        this.timeout(140 * 1000)
+
+        addFileAndCat(nodes[1], [nodes[2], nodes[3]], data, done)
       })
     })
 
-    it('three hops', (done) => {
+    describe('a Go bootstrap node in an hybrid land', () => {
+      let bootstrapAddr
+      const nodes = []
       const data = crypto.randomBytes(9001)
 
-      waterfall([
-        (cb) => goD3.api.add(data, cb),
-        (res, cb) => jsD.api.cat(res[0].hash, cb),
-        (stream, cb) => stream.pipe(bl(cb))
-      ], (err, file) => {
-        expect(err).to.not.exist()
-        expect(file).to.be.eql(data)
-        done()
+      // spawn bootstrap daemon and get address
+      before(function (done) {
+        this.timeout(60 * 1000)
+
+        spawnGoDaemon([], (err, node) => {
+          expect(err).to.not.exist()
+          nodes.push(node)
+
+          getBootstrapAddr(nodes[0], (err, res) => {
+            expect(err).to.not.exist()
+            bootstrapAddr = res
+            done()
+          })
+        })
+      })
+
+      // spawn daemons
+      before(function (done) {
+        this.timeout(70 * 1000)
+        const bootstrap = bootstrapAddr ? [bootstrapAddr] : []
+
+        parallel([
+          (cb) => spawnJsDaemon(bootstrap, cb),
+          (cb) => spawnGoDaemon(bootstrap, cb),
+          (cb) => spawnJsDaemon(bootstrap, cb)
+        ], (err, spawnedNodes) => {
+          expect(err).to.not.exist()
+          spawnedNodes.forEach((node) => nodes.push(node))
+
+          const testDhtReady = (cb) => {
+            nodes[0].api.swarm.peers((err, peers) => {
+              if (err) {
+                return cb(err)
+              }
+              cb(null, peers.length === 3)
+            })
+          }
+
+          // Wait for peers to have each other on the DHT
+          waitFor(testDhtReady, {
+            name: 'dht peers ready',
+            timeout: 30 * 1000,
+            interval: 5 * 1000
+          }, done)
+        })
+      })
+
+      after(function (done) {
+        this.timeout(80 * 1000)
+        each(nodes, (node, cb) => {
+          node.stop(cb)
+        }, done)
+      })
+
+      it('should get from the network after being added', function (done) {
+        this.timeout(60 * 1000)
+
+        addFileAndCat(nodes[1], [nodes[2], nodes[3]], data, done)
       })
     })
   })
```

Comment:
@vasco-santos I believe the tests above would be better if the tests were described through which nodes play which role (who is the node providing the content, who is the node fetching the content, who is the node bootstrapping the network, who are the other nodes) and make it just a simple config rather than hard wiring things by hand in every test.

What do you think?